File: app/src/test/java/com/matchthree/game/engine/MatchDetectorTest.kt
package com.matchthree.game.engine

import com.matchthree.game.Boards
import com.matchthree.game.model.Position
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class MatchDetectorTest {

    private val detector = MatchDetector

    @Test
    fun `horizontal run of 3 is found`() {
        val board = Boards.fromRows("RRR")
        val matches = detector.findMatches(board)
        assertEquals(1, matches.size)
        assertEquals(listOf(Position(0, 0), Position(0, 1), Position(0, 2)), matches[0].positions)
    }

    @Test
    fun `run of 4 is a single maximal match`() {
        val board = Boards.fromRows("RRRR")
        val matches = detector.findMatches(board)
        assertEquals(1, matches.size)
        assertEquals(4, matches[0].size)
        assertEquals(listOf(Position(0, 0), Position(0, 1), Position(0, 2), Position(0, 3)), matches[0].positions)
    }

    @Test
    fun `vertical run of 3 is found`() {
        val board = Boards.fromRows("R..", "R..", "R..")
        val matches = detector.findMatches(board)
        assertEquals(1, matches.size)
        assertEquals(listOf(Position(0, 0), Position(1, 0), Position(2, 0)), matches[0].positions)
    }

    @Test
    fun `L shape yields horizontal and vertical matches`() {
        val board = Boards.fromRows(
            "RRR",
            "..R",
            "..R",
        )
        val matches = detector.findMatches(board)
        assertEquals(2, matches.size)
        assertTrue(matches.any { it.positions.all { p -> p.row == 0 } })
        assertTrue(matches.any { it.positions.all { p -> p.col == 2 } })
    }

    @Test
    fun `two separate runs in one row are both found`() {
        val board = Boards.fromRows("RRRBBRRR")
        val matches = detector.findMatches(board)
        assertEquals(2, matches.size)
        assertEquals(3, matches[0].size)
        assertEquals(3, matches[1].size)
    }

    @Test
    fun `alternating board has no matches`() {
        val board = Boards.fromRows(
            "RYRYRYRYR",
            "YRYRYRYRY",
            "RYRYRYRYR",
        )
        assertEquals(0, detector.findMatches(board).size)
    }

    @Test
    fun `empty cells break runs`() {
        val board = Boards.fromRows("RR.RR")
        assertEquals(0, detector.findMatches(board).size)
    }
}

File: godot/Tests/MatchDetectorTest.cs
using MatchThree.Engine.Model;
using MatchThree.Engine.Rules;
using NUnit.Framework;

namespace MatchThree.Engine.Tests;

public class MatchDetectorTest
{
    [Test]
    public void HorizontalRunOf3IsFound()
    {
        var board = Boards.FromRows("RRR");
        var matches = MatchDetector.FindMatches(board);
        Assert.That(matches, Has.Count.EqualTo(1));
        Assert.That(matches[0].Positions, Is.EqualTo(new[]
        {
            new Position(0, 0), new Position(0, 1), new Position(0, 2),
        }));
    }

    [Test]
    public void RunOf4IsASingleMaximalMatch()
    {
        var board = Boards.FromRows("RRRR");
        var matches = MatchDetector.FindMatches(board);
        Assert.That(matches, Has.Count.EqualTo(1));
        Assert.That(matches[0].Size, Is.EqualTo(4));
        Assert.That(matches[0].Positions, Is.EqualTo(new[]
        {
            new Position(0, 0), new Position(0, 1), new Position(0, 2), new Position(0, 3),
        }));
    }

    [Test]
    public void VerticalRunOf3IsFound()
    {
        var board = Boards.FromRows("R..", "R..", "R..");
        var matches = MatchDetector.FindMatches(board);
        Assert.That(matches, Has.Count.EqualTo(1));
        Assert.That(matches[0].Positions, Is.EqualTo(new[]
        {
            new Position(0, 0), new Position(1, 0), new Position(2, 0),
        }));
    }

    [Test]
    public void LShapeYieldsHorizontalAndVerticalMatches()
    {
        var board = Boards.FromRows(
            "RRR",
            "..R",
            "..R");
        var matches = MatchDetector.FindMatches(board);
        Assert.That(matches, Has.Count.EqualTo(2));
        Assert.That(matches.Any(m => m.Positions.All(p => p.Row == 0)), Is.True);
        Assert.That(matches.Any(m => m.Positions.All(p => p.Col == 2)), Is.True);
    }

    [Test]
    public void TwoSeparateRunsInOneRowAreBothFound()
    {
        var board = Boards.FromRows("RRRBBRRR");
        var matches = MatchDetector.FindMatches(board);
        Assert.That(matches, Has.Count.EqualTo(2));
        Assert.That(matches[0].Size, Is.EqualTo(3));
        Assert.That(matches[1].Size, Is.EqualTo(3));
    }

    [Test]
    public void AlternatingBoardHasNoMatches()
    {
        var board = Boards.FromRows(
            "RYRYRYRYR",
            "YRYRYRYRY",
            "RYRYRYRYR");
        Assert.That(MatchDetector.FindMatches(board), Is.Empty);
    }

    [Test]
    public void EmptyCellsBreakRuns()
    {
        var board = Boards.FromRows("RR.RR");
        Assert.That(MatchDetector.FindMatches(board), Is.Empty);
    }
}